### DISCUSSION TITLE
Check if one-file or one-dir

### DISCUSSION THREAD

**Discussion by dimitriepirghie** [pyinstaller/pyinstaller#5946] [PyInstaller] [answered]
Hello,

There is a way to check if the application is running on `one-file` or `one-dir` mode ?
I am using `pyupdater` for updates and now I want to provide also `one-dir` builds and I would need to have this information for  choosing the right endpoint :D.

Thanks.

**Comment by bwoodsend** [comment, marked as answer]
I guess you could compare `sys._MEIPASS` and `sys.executable`. If you're in `onedir` mode then `sys._MEIPASS == os.path.dirname(sys.executable)`. Does feel like a bit of a hack though.

```python
# test.py
import sys

print(sys.executable)
print(sys._MEIPASS)
```

```bash
> pyinstaller test.py
> ./dist/test/test
/e/notebooks/pyi-issues/5946/dist/test/test
/e/notebooks/pyi-issues/5946/dist/test

> pyinstaller --onefile --name=onefile test.py
> ./dist/onefile 
/e/notebooks/pyi-issues/5946/dist/onefile
/tmp/_MEIOmfMMX
```

**Reply by dimitriepirghie** [reply]
Thank you. I am attribute or bool returning function will be the best but for the moment it's ok like this.

**Comment by Phxntxm** [comment]
Just a small comment on this if this is the proper place for it, it would be nice if there was a supported way to do this that's not a "hack"

My use case: I want to log information in case something goes wrong so users can provide me these logs. For dynamic uses (I'll be providing ways to build themselves, if they really want to) I want to then choose where the logging goes. If we're in onedir mode, then I can simply log there... if not though, that's the temp dir and I can't use it for logging because the log file will go bye bye too. So I need to know which mode I am so that I can choose where to put the log file. I can't dynamically just use `sys.executable` either because putting a log file in a place like `/usr/bin/` is definitely wrong (but would be fine for windows) etc

**Reply by rokm** [reply]
FWIW, I think in the presented scenario, the logs should go into application-specific sub-directory in the user's home directory (along with any user-specific configuration and/or application state). Then, it doesn't matter if you're in `onedir` or in `onefile` mode (or in a  `onedir` application that was `MERGE`'d with another), nor does it matter how the onedir dist was "installed" (e.g., placed in `/opt/<app.dir>` with root as owner and no write permissions for regular users), or how many users are using it, etc. Plus, your users will always know where to look for the logs.

**Reply by Phxntxm** [reply]
Mmm thank you for the insight, at first all I was really doing is placing a single config file in `~/.config/app` and providing a binary that can be placed whenever. But I think it may be a good idea to do a proper installer, an application directory like you mentioned instead. I've just always been hesitant about crowding someone's $HOME for some reason...

**Reply by bwoodsend** [reply]
I'd still agree that we should have a flag set somewhere accessible (and documented). But as @rokm says, for your case I'd generally recommend not writing into the onedir's folder because it's not uncommon for security reasons to *lock* executable locations after installing so that they require sudo permissions to edit. Putting logs in there would then make that impossible without running the application as sudo user.
